src/components/Web3SendTransactionButton.tsx

```diff
@@ -1,54 +1,73 @@
 import { Button, CircularProgress, Tooltip } from "@material-ui/core";
 import DoneIcon from "@material-ui/icons/Done";
 import ErrorIcon from "@material-ui/icons/Error";
+import TimerIcon from "@material-ui/icons/Timer";
+import moment from "moment";
 import React, { useState } from "react";
 import { Transaction } from "web3-core/types";
+import { now } from "../time";
 import MetamaskIcon from "./MetamaskIcon";
 import NoWeb3Tooltip from "./NoWeb3Tooltip";
 import { useWeb3 } from "./Web3Context";
 
 interface Props {
   transaction: Transaction;
-  actionReady: boolean;
+  minTimestamp?: number;
 }
 
 enum TransactionState {
   Initial,
   Signing,
   Sent,
-  Error
+  Error,
+  Unavailable
 }
 
-function Web3SendTransactionButton({ transaction, actionReady }: Props) {
+function Web3SendTransactionButton({ transaction, minTimestamp }: Props) {
   const web3 = useWeb3();
   const [state, setState] = useState(TransactionState.Initial);
 
+  const whenReadyMessage =
+    !minTimestamp || minTimestamp <= now()
+      ? "now"
+      : moment.unix(minTimestamp).fromNow();
+
   const onClickHandler = web3
     ? async () => {
-        setState(TransactionState.Signing);
-        try {
-          // Have to fetch default account ourselves until this works:
-          // https://github.com/MetaMask/metamask-extension/issues/6339
-          const accounts = await web3.eth.getAccounts();
-          await web3.eth.sendTransaction({
-            ...transaction,
-            from: accounts[0]
-          });
-          setState(TransactionState.Sent);
-        } catch (e) {
-          setState(TransactionState.Error);
+        const actionReady =
+          !minTimestamp ||
+          (!!web3 &&
+            (await web3.eth.getBlockNumber().then(
+              async n =>
+                await web3.eth.getBlock(n).then(block => {
```

Comment:
:wrench: Two improvements here:

1. You can do `getBlock("latest")` to avoid doing two calls: https://web3js.readthedocs.io/en/1.0/web3-eth.html#getblock
2. If you use `await` anyway, you can do whatever you do inside the `then` block outside by assigning it to a variable:

```js
let block = await web3.eth.getBlock("latest");
```

Reply:
Much cleaner this way!